hal/sparcv8leon/gaisler/timer.c

```diff
@@ -75,40 +93,64 @@ static inline void timer_setReloadValue(int timer, u32 val)
 }
 
 
-static void timer_setPrescaler(int timer, u32 freq)
+static void timer_setPrescaler(int timer, u32 interval)
 {
 	u32 prescaler = SYSCLK_FREQ / 1000000; /* 1 MHz */
-	u32 ticks = (SYSCLK_FREQ / prescaler) / freq;
 
-	timer_setReloadValue(timer, ticks - 1);
+	timer_setReloadValue(timer, interval - 1);
 	*(timer_common.timer0_base + GPT_SRELOAD) = prescaler - 1;
 
-	timer_common.ticksPerFreq = ticks;
+	timer_common.ticksPerFreq = interval;
 }
 
 
 time_t hal_timerGetUs(void)
 {
+	u32 regVal;
 	time_t val;
 	spinlock_ctx_t sc;
 
 	hal_spinlockSet(&timer_common.sp, &sc);
+
+	regVal = *(timer_common.timer0_base + GPT_TCNTVAL(TIMER_TIMEBASE));
+
+	/* Check if there's pending irq */
+	if ((*(timer_common.timer0_base + GPT_TCTRL(TIMER_TIMEBASE)) & TIMER_INT_PENDING) != 0) {
+		++timer_common.jiffies;
+		timer_clearIrq(TIMER_TIMEBASE);
+		/* Timer might've just wrapped-around, take counter value again */
+		regVal = *(timer_common.timer0_base + GPT_TCNTVAL(TIMER_TIMEBASE));
+	}
 	val = timer_common.jiffies;
+
 	hal_spinlockClear(&timer_common.sp, &sc);
 
-	return val * 1000ULL;
+	return val * timer_common.ticksPerFreq + timer_common.ticksPerFreq - regVal;
 }
 
 
 void hal_timerSetWakeup(u32 waitUs)
 {
+	spinlock_ctx_t sc;
+
+	hal_spinlockSet(&timer_common.sp, &sc);
+
+	/* Disable timer */
+	*(timer_common.timer0_base + GPT_TCTRL(TIMER_WAKEUP)) = 0;
+	timer_clearIrq(TIMER_WAKEUP);
+
+	/* Configure one shot timer */
+	timer_setReloadValue(TIMER_WAKEUP, waitUs);
+	*(timer_common.timer0_base + GPT_TCTRL(TIMER_WAKEUP)) = TIMER_ENABLE | TIMER_INT_ENABLE | TIMER_LOAD;
```

Comment:
Just making sure - `TIMER_LOAD` flag cause reload value to be transferred to the counter register, so the timer starts with the reload value, not the old value of the counter?